### PR TITLE
Extended request spans

### PR DESCRIPTION
**Pull request by piodul**
Provides more information in trace spans emitted for unprepared query, prepared query and batch requests.

The following information is provided:

- `kind`: type of operation (`unprepared`, `prepared`, `batch`),
- `contents` - the query string (applicable only to unprepared queries),
- `routing_key`: routing key of the request (if known, i.e. applies only to prepared queries),
- `token`: partition token relevant to the request (if known, i.e. applies only to prepared queries),
- `request_size`: serialized size of query parameters (not implemented for batches),
- `result_size`: serialized size of rows returned in response,
- `result_rows`: number of rows returned in response,
- `shard`: ID of the shard the request was served on,
- `speculative_executions` - the number of times this request was speculatively executed,

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] ~I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~I added appropriate `Fixes:` annotations to PR description.~

### CODE REVIEW

File: scylla/src/utils/pretty.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+use std::fmt::{LowerHex, UpperHex};
+
+pub(crate) struct HexBytes<'a>(pub &'a [u8]);
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/utils/pretty.rs:3]: Yeah, I feel the same about `pub`s. I missed this one, thanks, I'll change it to `pub(crate)`.

File: scylla/src/transport/session.rs [hunk 2]
```diff
@@ -1753,11 +1753,11 @@ impl Session {
...
-    pub fn calculate_token(
+    fn calculate_token_and_routing_key(
         &self,
         prepared: &PreparedStatement,
         serialized_values: &SerializedValues,
-    ) -> Result<Option<Token>, QueryError> {
+    ) -> Result<Option<(Token, Bytes)>, QueryError> {
         if !prepared.is_token_aware() {
             return Ok(None);
         }
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:1799]: I saw that other drivers used the "routing key" term to indicate a "partition key" serialized in a particular form. I see that we are using the "partition key" term across the existing code in the driver, so I'll change.

File: scylla/src/transport/session.rs [hunk 3]
```diff
@@ -1766,7 +1766,19 @@ impl Session {
...
+    pub fn calculate_token(
+        &self,
+        prepared: &PreparedStatement,
+        serialized_values: &SerializedValues,
+    ) -> Result<Option<Token>, QueryError> {
+        match self.calculate_token_and_routing_key(prepared, serialized_values) {
+            Ok(Some((token, _))) => Ok(Some(token)),
+            Ok(None) => Ok(None),
+            Err(err) => Err(err),
+        }
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs]: I'd rather keep `calculate_token` as it is - a user recently sent a PR to make it public, so I assume they would like to keep this interface.

File: scylla/src/transport/cluster.rs [hunk 4]
```diff
@@ -437,6 +437,16 @@ impl ClusterData {
...
+        self.get_token_endpoints_iter(keyspace, token)
+            .cloned()
+            .collect()
+    }
+
+    pub(crate) fn get_token_endpoints_iter(
+        &self,
+        keyspace: &str,
+        token: Token,
+    ) -> impl Iterator<Item = &Arc<Node>> {
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/cluster.rs:449]: In that case, I'll just change `get_token_endpoints` to return an iterator. It's quite easy for users to collect to a Vec, and the less methods in the interface, the better.

File: scylla/src/transport/session.rs [hunk 5]
```diff
@@ -899,10 +903,18 @@ impl Session {
...
-            prepared_id = %format_args!("{:X}", prepared.get_id())
-        );
+        let span = RequestSpan::new_prepared(token_and_key.as_ref(), serialized_values.size());
+
+        if !span.span().is_disabled() {
+            if let (Some(keyspace), Some(token)) = (statement_info.keyspace.as_ref(), token) {
+                let cluster_data = self.get_cluster_data();
+                let replicas: smallvec::SmallVec<[_; 8]> = cluster_data
+                    .get_token_endpoints_iter(keyspace, token)
+                    .collect();
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:913]: Just a wild guess. Sounded high enough to be sufficient in the most cases. If you have a more informed guess then I'll gladly change.

File: scylla/src/transport/session.rs [hunk 6]
```diff
@@ -1895,3 +1900,155 @@ impl<'a> ExecuteQueryContext<'a> {
...
+    ) -> Self {
+        use crate::utils::pretty::HexBytes;
+        use tracing::field::Empty;
+
+        let span = trace_span!(
+            "Request",
+            kind = "prepared",
+            routing_key = Empty,
+            token = Empty,
+            //
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:1976]: I just wanted to separate the common fields (at the end) from those that are specific to the span type.

File: scylla/src/transport/session.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/src/transport/session.rs]: Let me rephrase: the interface I propose is as follows:
```
pub fn calculate_token(
    &self,
    prepared: &PreparedStatement,
    serialized_values: &SerializedValues,
) -> Result<Option<Token>, QueryError> {
```
(unchanged)

and:
```
fn calculate_routing_key(
    &self,
    prepared: &PreparedStatement,
    serialized_values: &SerializedValues,
) -> Result<Option<Token>, QueryError> {
```

with `calculate_token` implementation as follows:
```
pub fn calculate_token(
    &self,
    prepared: &PreparedStatement,
    serialized_values: &SerializedValues,
) -> Result<Option<Token>, QueryError> {
    match self.calculate_routing_key(prepared, serialized_values) {
        Ok(Some(routing_key)) => {
             let partitioner_name = prepared.get_partitioner_name();
             Ok(Some(partitioner_name.hash(&routing_key)))
        }
        otherwise => otherwise,
    }
    ```

File: scylla/src/transport/session.rs [hunk 7]
```diff
@@ -899,10 +903,18 @@ impl Session {
             is_confirmed_lwt: prepared.is_confirmed_lwt(),
         };
 
-        let span = trace_span!(
-            "Request",
-            prepared_id = %format_args!("{:X}", prepared.get_id())
-        );
+        let span = RequestSpan::new_prepared(token_and_key.as_ref(), serialized_values.size());
+
+        if !span.span().is_disabled() {
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:908]: 1. It's to record the list of replicas. This information cannot be computed for an unprepared query.
2. AFAIK, whether a span is enabled or not depends on how the user configures the tracing library. For example, you can disable all traces below given verbosity level, or provide a custom filter that analyzes the trace field names etc. This is just a small optimization.

File: scylla/src/transport/session.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs]: Ok, makes sense.

File: scylla/src/transport/session.rs [hunk 8]
```diff
@@ -1895,3 +1900,155 @@ impl<'a> ExecuteQueryContext<'a> {
...
+            span: tracing::Span::none(),
+            speculative_executions: 0.into(),
+        }
+    }
+
+    pub(crate) fn record_shard_id(&self, conn: &Connection) {
+        if let Some(info) = conn.get_shard_info() {
+            self.span.record("shard", info.shard);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs:2031]: Nope, `info.shard` is a `u16`.

**Comment by piodul** [comment]
v1.1: addressed most of @wprzytula 's remarks.

We also discussed offline that query_iter traces need improvement - they don't really work, and data could be recorded better. I'll try to address those issues in the next version, coming soon.

File: scylla/src/transport/session.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply scylla/src/transport/session.rs]: Again about 1.: you do the check for prepared statement only, because only there you record replicas. And only recording replicas is considered by you to be worth such an optimisation. Correct?

[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/session.rs]: Yes.

**Comment by piodul** [comment]
v2:
- Made sure that the worker future created in query_iter/execute_iter will use their parent span
- Got rid of some invalid "Request" spans that were previously included in query_iter/execute_iter
- Sprinkled `with_current_subscriber` across the code so that tracing subscriber propagates properly

**Review by cvybhu — APPROVED** [review]
LGTM
